contrib/babelfishpg_tsql/src/tsqlIface.cpp

```diff
@@ -1814,65 +1885,125 @@ class tsqlBuilder : public tsqlCommonMutator
+				
+				// Change double-quoted string to single-quoted string:
+				str = rewriteDoubleQuotedString(str);	
+				size_t startPosition = ctx->start->getStartIndex();
+				rewritten_query_fragment.emplace(std::make_pair(startPosition, std::make_pair(::getFullText(ctx), str)));
+			}
+			else
+			{
+				// This is a single-quoted string, no further action needed
+				Assert(str.front() == '\'');
+				Assert(str.back() == '\'');		    						
+			}
+		}
+	}
+
 	void exitExecute_parameter(TSqlParser::Execute_parameterContext *ctx) override
 	{
-		if (ctx->constant() || ctx->id())
+		if (ctx->id())
 		{
-			// BABEL-2395, BABEL-3640: embedded single quotes are allowed in procedure parameters.
-			// in tsqlMutator::enterExecute_parameter, we replace "" with '', and also record the place of "" into 
-			// double_quota_places, then in here, we'll replace all ' into '' if it's inside ""
-			// we replace " in tsqlMutator::enterExecute_parameter first to pass the syntax validation
-			// then we'll use the double quota rule to replace "'A'" to '''A'''
-			if (pg_strcasecmp(getProcNameFromExecParam(ctx).c_str(), "sp_tables") == 0)
+			// A stored procedure parameter which is parsed as a column name (= identifier)
+			// is either an unquoted string, or a double-quoted string.
+			// For a procedure call parameter, a double-quoted string is interpreted 
+			// as a string even with QUOTED_IDENTIFIER=ON.
+			
+			std::string str = getFullText(ctx->id());
+
+			if (str.front() == '"')
 			{
-				std::string argStr;
-				if (ctx->constant())
-					argStr = getFullText(ctx->constant());
-				else if (ctx->id())
-					argStr = getFullText(ctx->id());
-
-				std::string newStr;
-				if (argStr.size() > 1 && argStr.front() == '\'' && argStr.back() == '\''
-					&& std::binary_search (double_quota_places.begin(), double_quota_places.end(), parameterIndex))
-				{
-					newStr += '\'';
-					for (std::string::iterator iter = argStr.begin() + 1; iter != argStr.end() - 1; ++iter)
-					if (*iter == '\'')
-					{
-						newStr += "\'\'";
-					} else{
-						newStr += *iter;
-					}
-					newStr += '\'';
-					if (ctx->constant())
-						rewritten_query_fragment.emplace(std::make_pair(ctx->start->getStartIndex(),
-						 	std::make_pair(::getFullText(ctx->constant()), newStr)));
-					else 
-						rewritten_query_fragment.emplace(std::make_pair(ctx->start->getStartIndex(), std::make_pair(getFullText(ctx->id()), newStr)));
-				}
+				Assert(str.back() == '"');
+								
+				// Change double-quoted string to single-quoted string
+				str = rewriteDoubleQuotedString(str);		    	
+				rewritten_query_fragment.emplace(std::make_pair(ctx->start->getStartIndex(), std::make_pair(getFullText(ctx->id()), str)));
+			}
+			else if (str.front() == '\'')
+			{
+				// This is a single-quoted string, no further action needed	
+				Assert(str.back() == '\'');		    						
+			}
+			else 
+			{
+				// This is an unquoted string parameter which has been parsed as a column name.
```

Comment:
No need to keep an empty 
{}
If nothing need to be done here

Reply:
I'd like to keep it as it documents the logic in the code.